csharp/src/Ice/ACM.cs

```diff
@@ -2,98 +2,155 @@
-                Timeout = defaults.Timeout;
-            }
+        internal AcmConfig(Communicator communicator, string prefix, AcmConfig defaults)
+        {
+            Timeout = communicator.GetPropertyAsTimeSpan($"{prefix}.Timeout") ?? defaults.Timeout;
 
-            int hb = communicator.GetPropertyAsInt($"{prefix}.Heartbeat") ?? (int)defaults.Heartbeat;
-            if (hb >= (int)ACMHeartbeat.HeartbeatOff && hb <= (int)ACMHeartbeat.HeartbeatAlways)
+            if (communicator.GetPropertyAsInt($"{prefix}.Heartbeat") is int hearbeat)
             {
-                Heartbeat = (ACMHeartbeat)hb;
+                if (hearbeat < (int)AcmHeartbeat.HeartbeatOff || hearbeat > (int)AcmHeartbeat.HeartbeatAlways)
+                {
+                    throw new InvalidConfigurationException($"invalid value for property `{prefix}.Heartbeat'");
+                }
+                Heartbeat = (AcmHeartbeat)hearbeat;
             }
             else
             {
-                logger.Warning($"invalid value for property `{prefix}.Heartbeat', default value will be used instead");
                 Heartbeat = defaults.Heartbeat;
             }
 
-            int cl = communicator.GetPropertyAsInt($"{prefix}.Close") ?? (int)defaults.Close;
-            if (cl >= (int)ACMClose.CloseOff && cl <= (int)ACMClose.CloseOnIdleForceful)
+            if (communicator.GetPropertyAsInt($"{prefix}.Close") is int close)
             {
-                Close = (ACMClose)cl;
+                if (close < (int)AcmClose.CloseOff || close > (int)AcmClose.CloseOnIdleForceful)
+                {
+                    throw new InvalidConfigurationException($"invalid value for property `{ prefix}.Close'");
+                }
+                Close = (AcmClose)close;
             }
             else
             {
-                logger.Warning($"invalid value for property `{prefix}.Close', default value will be used instead");
                 Close = defaults.Close;
             }
         }
-
-        public ACMConfig Clone() => (ACMConfig)MemberwiseClone();
-
-        public int Timeout;
-        public ACMHeartbeat Heartbeat;
-        public ACMClose Close;
     }
 
-    public interface IACMMonitor : ITimerTask
+    internal class ConnectionFactoryAcmMonitor : IAcmMonitor
     {
-        void Add(Connection con);
-        void Remove(Connection con);
-        void Reap(Connection con);
+        public Acm Acm => new Acm(_config.Timeout, _config.Close, _config.Heartbeat);
 
-        IACMMonitor Acm(int? timeout, ACMClose? c, ACMHeartbeat? h);
-        ACM GetACM();
-    }
+        private readonly List<(Connection Connection, bool Remove)> _changes =
+            new List<(Connection Connection, bool Remove)>();
+        private Communicator? _communicator;
+        private readonly AcmConfig _config;
+        private readonly HashSet<Connection> _connections = new HashSet<Connection>();
```

Comment:
An alternative option to _changes and the _connections HashSet would be to use a `ConcurrentDictionary<Connection, bool>`. The bool value would be ignored (we use the dictionary as a HashSet). 

One issue with this however is that the RunTimerTask iteration on the concurrent dictionary might return removed connections and call Monitor on them... it's something however the Monitor implementation could deal with.

It's also unclear whether or not this would be more efficient or not... perhaps not if iterating over the enumeration isn't lock free.

So... hard to say wether or not it would be better without deeper knowledge on the ConcurrentDictionary implementation. What's important here is that we don't copy the _connection for each iteration since we might be dealing with a large number of connections with servers that accept thousands of connections.

Reply:
Seems iterating is lock free, what will be the benefices of this, that we don't need to hold the lock for iterating the collection?

Reply:
Keeped the hashset for now, if we want to implement the monitor in the factory we can probably get rid of this class

Reply:
Agreed, let's revisit this once we get rid of the reaping in the monitor.